KNN 59
======

Adrián Arroyo Calle - TAA Marzo 2019




En esta entrega tenemos un conjunto de datos formado por imágenes en escala de grises que representan cincos y nueves de forma manuscrita. Usando estos datos analizaremos el comportamiento del clasificador perezoso K-NN, tratando de encontrar el K óptimo. Para ello se usa Python con el módulo sklearn.neighbors.

In [1]:
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import KFold

In [2]:
entrenae = pd.read_csv("entrenae59.csv")
entrenae_x = entrenae.drop("digito",axis=1)
entrenae_y = entrenae["digito"]

In [3]:
pruebae = pd.read_csv("pruebae59.csv")
pruebae_x = pruebae.drop("digito",axis=1)
pruebae_y = pruebae["digito"]

In [4]:
res = dict()
cv = KFold(10)
for k in range(1,11,2):
    c = KNeighborsClassifier(n_neighbors=k)
    i = 0
    acumula = 0
    cv_iter = cv.split(entrenae_x,entrenae_y)
    for train, test in cv_iter:
        c.fit(entrenae_x.iloc[train],entrenae_y.iloc[train])
        aciertos = c.score(entrenae_x.iloc[test],entrenae_y.iloc[test])
        i += 1
        acumula += aciertos
    print("K=%d Tasa acierto: %f" % (k,acumula/i))

K=1 Tasa acierto: 0.901180
K=3 Tasa acierto: 0.901938
K=5 Tasa acierto: 0.964698
K=7 Tasa acierto: 0.963922
K=9 Tasa acierto: 0.960857


In [5]:
res = dict()
for k in range(1,11,2):
    c = KNeighborsClassifier(n_neighbors=k)
    c.fit(entrenae_x,entrenae_y)
    res[k] = c.score(pruebae_x,pruebae_y)
res

{1: 0.8950617283950617,
 3: 0.8950617283950617,
 5: 0.8966049382716049,
 7: 0.8935185185185185,
 9: 0.8935185185185185}

In [8]:
import math
math.sqrt((res[5]*(1-res[5]))/pruebae_y.shape[0])

0.011960875193690374

En ambos casos se obtiene que el K óptimo es 5.

Su tasa de error con el conjunto de entrenamiento es 0.8966 con una desviación estándar de 0.0119